test-adapter/src/main/scala/org/scalajs/sbttestadapter/RunnerAdapter.scala

```diff
@@ -0,0 +1,120 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js test adapter      **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2017, LAMP/EPFL        **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package org.scalajs.testadapter
+
+import scala.collection.concurrent.TrieMap
+
+import scala.concurrent._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+
+import scala.util._
+
+import org.scalajs.testcommon._
+
+import sbt.testing._
+
+import FutureUtil._
+import TestAdapter.ManagedRunner
+
+private[scalajs] final class RunnerAdapter private (runnerArgs: RunnerArgs,
+    master: ManagedRunner, testAdapter: TestAdapter) extends Runner {
+
+  private val runId = runnerArgs.runId
+  private val rpcGetter = () => getRunnerRPC()
+
+  private val slaves = TrieMap.empty[Long, ManagedRunner]
+
+  // Route master messages to slaves.
+  master.mux.attach(JVMEndpoints.msgMaster, runId) { msg =>
+    slaves(msg.slaveId).mux.send(JSEndpoints.msgSlave, runId)(msg.msg)
+  }
+
+  def args(): Array[String] = runnerArgs.args.toArray
+
+  def remoteArgs(): Array[String] = runnerArgs.remoteArgs.toArray
+
+  def tasks(taskDefs: Array[TaskDef]): Array[Task] = {
+    val rpc = getRunnerRPC()
+    val taskInfos = Await.result(
+        rpc.call(JSEndpoints.tasks, runId)(taskDefs.toList),
+        testAdapter.timeout)
+
+    taskInfos.map(new TaskAdapter(_, runId, rpcGetter)).toArray
+  }
+
+  def done(): String = synchronized {
+    def callDone(mRunner: ManagedRunner): Future[Try[String]] =
+      mRunner.mux.call(JSEndpoints.done, runId)(()).liftToTry
+
+    val futureSummary = for {
+      slaveTries <- Future.sequence(slaves.values.map(callDone))
+      summaryTry <- callDone(master)
+    } yield {
+      slaveTries.foreach(_.get)
+      summaryTry.get
+    }
+
+    try {
+      // Double timeout since we need two roundtrips (slaves and then master).
+      Await.result(futureSummary, testAdapter.timeout * 2)
+    } finally {
+      slaves.values.foreach(_.mux.detach(JVMEndpoints.msgSlave, runId))
+      master.mux.detach(JVMEndpoints.msgMaster, runId)
+
+      slaves.clear()
+      testAdapter.runDone(runId)
+    }
+  }
+
+  private def getRunnerRPC(): RunMuxRPC = {
+    val mRunner = testAdapter.getRunnerForThread()
+
+    if (mRunner != master && !slaves.contains(mRunner.id)) {
```

Comment:
This actually looks like it will never call `JSEndpoints.createSlaveRunner` if we happen to call this method from the thread of the master, even if we intend to use that `RunMuxRPC` for slave tasks.